packages/graphql-codegen-near-operation-file-preset-with-resolver-metadata/src/definitions-metadata.ts

```diff
@@ -0,0 +1,364 @@
+    }
+
+    output.schemaMetadata = mergeSchemaDefinitions({ types: {} }, [
+      ...Object.values(operationDefinitions).map((item) => item.definitions),
+      ...Object.values(usedFragmentDefinitions).map((item) => item.definitions),
+    ]);
+
+    return output;
+  }
+
+  return null;
+}
+
+const TypeKind = {
+  SCALAR: 1,
+  OBJECT: 2,
+  INTERFACE: 3,
+  UNION: 4,
+  ENUM: 5,
+  INPUT: 6,
+} as const;
+
+function getResolverMetadata(
+  operationDefinitions: Record<string, ExtractMinimalViableSchemaResult>,
+  fragmentDefinitions: Record<string, ExtractMinimalViableSchemaResult>,
+) {
+  const resolverMetadata: Record<string, string[]> = {};
+
+  const interfaceFields: Record<string, string[]> = {};
+  for (const [, { definitions }] of Object.entries(operationDefinitions)) {
+    for (const [type, [typeKind, fields]] of Object.entries(
+      definitions.types,
+    )) {
+      if (typeKind === TypeKind.INTERFACE) {
+        for (const field of Object.keys(fields ?? {})) {
+          interfaceFields[type] ??= [];
+          if (!interfaceFields[type].includes(field)) {
+            interfaceFields[type].push(field);
+          }
+        }
+      }
+    }
+  }
+
+  const convertDefinitionsToResolverMetadataFormat = (
+    inputDefinitions: Record<string, ExtractMinimalViableSchemaResult>,
+  ) => {
+    for (const [, { definitions }] of Object.entries(inputDefinitions)) {
+      for (const [type, [typeKind, fields, interfaces]] of Object.entries(
+        definitions.types,
+      )) {
+        if (typeKind === TypeKind.ENUM) {
+          continue;
+        }
+        if (typeKind === TypeKind.UNION) {
+          resolverMetadata[type] ??= [];
+          if (!resolverMetadata[type].includes("__resolveType")) {
+            resolverMetadata[type].push("__resolveType");
+          }
+          continue;
+        }
+
+        if (typeKind === TypeKind.INTERFACE) {
+          resolverMetadata[type] ??= [];
+          if (!resolverMetadata[type].includes("__resolveType")) {
+            resolverMetadata[type].push("__resolveType");
+          }
+          continue;
+        }
+
+        for (const field of Object.keys(fields ?? {})) {
+          resolverMetadata[type] ??= [];
+          if (!resolverMetadata[type].includes(field)) {
+            resolverMetadata[type].push(field);
+          }
+        }
+
+        if (interfaces?.length) {
+          for (const interfaceName of interfaces) {
+            const fields = interfaceFields[interfaceName];
+            if (!fields) {
+              continue;
+            }
+            for (const field of fields) {
+              resolverMetadata[type] ??= [];
+              if (!resolverMetadata[type].includes(field)) {
+                resolverMetadata[type].push(field);
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  convertDefinitionsToResolverMetadataFormat(operationDefinitions);
+  convertDefinitionsToResolverMetadataFormat(fragmentDefinitions);
```

Comment:
I don't fully follow why we do this separately on operations and fragments? Different operations and fragments have many repeating fields. Can we run it on merged schema when everything is already deduplicated by the merge itself?